### PR TITLE
perf: use host UUID map for O(1) tablet replica lookups in Pick

### PR DESCRIPTION
**Pull request by mykaul**
## Summary

- Replace O(replicas × hosts) linear scan with O(1) map lookup when resolving tablet replica UUIDs to `*HostInfo` in `tokenAwareHostPolicy.Pick()`
- Add `hostsByID` atomic map field, rebuilt on host add/remove under existing lock
- On clusters with 100+ hosts and 3 tablet replicas, this reduces ~300 UUID comparisons per query to 3 map lookups

## Benchmark results

Confirms constant time regardless of cluster size:

| Hosts | ns/op | allocs |
|-------|-------|--------|
| 3     | 424   | 9      |
| 10    | 379   | 9      |
| 50    | 413   | 9      |
| 100   | 389   | 9      |
| 200   | 378   | 9      |

---
> **Maintenance note (rebased onto `master`):** This branch was rebased onto the latest `master` (tip `bed2bb09`) and all unit tests pass (only the known SSL/cloud cert env failures, unrelated to this change).
> The benchmark figures above were measured **pre-rebase** against the branch's original merge-base; a re-benchmark against current `master` is pending and numbers will be refreshed.

### CODE REVIEW

**Comment by mykaul** [comment]
@coderabbitai full review

File: policies.go [hunk 1]
```diff
@@ -626,11 +629,29 @@ func (t *tokenAwareHostPolicy) SetPartitioner(partitioner string) {
...
+	for _, h := range hosts {
+		m[h.hostId] = h
+	}
+	t.hostsByID.Store(m)
+}
+
+// getHostsByID returns the current host-by-UUID map for lock-free lookups.
+func (t *tokenAwareHostPolicy) getHostsByID() map[UUID]*HostInfo {
+	m, _ := t.hostsByID.Load().(map[UUID]*HostInfo)
+	return m
```
[the remark this answers is not in the document]
Comment (mykaul) [reply policies.go:645]: My AI disagrees with your AI:
The two stores happen back-to-back under t.mu, so the window is sub-microsecond. The only observable effect is that a concurrent Pick might briefly miss a just-added host (same as pre-optimization behavior) or briefly resolve a just-removed host (handled gracefully by the connection layer). The snapshot struct approach would add substantial complexity for no practical benefit — topology changes are already eventually-consistent operations.

**Comment by mykaul** [comment]
## Review: O(1) tablet replica lookup (`hostsByID` map)

Deep review of the `tokenAwareHostPolicy.Pick()` hot-path change. Reviewed in an isolated worktree against a fresh rebase on `origin/master`.

### Findings

| Severity | Area | Finding |
|----------|------|---------|
| — (PASS) | O(1) correctness | `hostMap[UUID(replica.HostUUIDValue())]` + `if host != nil` is semantically equivalent to the prior `O(replicas×hosts)` linear scan: match → append, miss → skip. No behavioral regression. |
| — (PASS) | nil-map deref | `getHostsByID()` returns a `nil` map when `hostsByID` was never stored; reading from a nil map is safe Go (miss, not panic). |
| — (PASS) | lifecycle / race | `hostsByID` is always **replaced** with a freshly built map via `atomic.Value.Store` (never mutated in place), under `t.mu`, on every path that mutates `t.hosts` (`AddHost`/`AddHosts`/`RemoveHost`). Read lock-free in `Pick`. `-race` confirms no concurrent map read/write. |
| — (PASS) | mixed-snapshot window (coderabbitai) | Re-verified the declined feedback. The two independent stores (`hostsByID`, `metadata`) under `t.mu`, read lock-free in `Pick`, can momentarily diverge (sub-µs). The connection layer already tolerates missed-add / stale-remove, matching the existing eventually-consistent COW design. The map-miss path is handled identically to the old scan. **Reasoning holds — no snapshot struct introduced.** |
| Low (addressed via tests) | edge-case coverage | The only test was the happy path. Added correctness + race tests for the review-focus gaps (stale/unknown UUID, partial miss, post-remove, empty map, concurrent rebuild). |

No HIGH/MEDIUM issues found. **No hot-path edits were made** — the `Pick()`/`hostsByID` logic is unchanged from the author's commit.

### Changes made (test-only)

Added `policies_test.go` coverage (build tag `unit`):
- `TestTokenAwareTabletReplicaMapMiss` — unknown replica UUID → skip, fall back to token ring, no panic / no nil host.
- `TestTokenAwareTabletReplicaPartialMiss` — mixed known/unknown replicas: known resolved, unknown skipped (presence/absence, not order — tablet replica order is not builder-guaranteed).
- `TestTokenAwareTabletReplicaAfterRemove` — stale UUID no longer resolved after `RemoveHost`; safe fallback.
- `TestTokenAwareTabletReplicaEmptyMap` — nil/empty `hostsByID` lookup is safe.
- `TestTokenAwareHostsByIDConcurrentRebuild` — concurrent `AddHost`/`RemoveHost` rebuilds vs lock-free `Pick` reads (`-race`).

### Test results

- `go build ./...`, `go vet ./...`: clean.
- `go test -count=1 -tags unit ./...`: **10838 passed, 233 skipped, 4 failed** — all 4 are the known PKI env failures (`TestSSLSimple`, `TestSSLSimpleNoClientCert`, scyllacloud cert tests) from missing `testdata/pki/ca.crt`. Not regressions.
- `go test -race -count=1 -tags unit .` (root pkg): **581 passed, 2 skipped, 11 failed** — all 11 are PKI/TLS env failures (`Failed parsing or appending certs` + their `with_TLS` parent subtests). **Zero `DATA RACE` reports.** New tests all pass under `-race`.

### Rebase

Cleanly rebased on `origin/master` (no conflicts).

### Push decision: **NOT pushed**

Per the hot-path policy: this PR changes `Pick()`/`hostsByID` correctness logic, which requires benchmarking before push, and the benchmark machine is busy. Commits are kept locally only. The author's commit is unchanged; my addition is test-only and sits on top.

### Benchmark recommendation

Existing benchmarks already cover this path (added in the prior UUID PR):

```
go test -tags unit -run='^$' -bench='BenchmarkTabletAwarePick' -benchmem .
go test -tags unit -run='^$' -bench='BenchmarkTabletAwarePickAllReplicas' -benchmem .
```

Recommend capturing `-count=10` before/after (`benchstat`) for `Hosts10/50/100`, focusing on the `Hosts100` case where the old `O(RF×H)=300`-comparison scan is replaced by `RF` map lookups — that is where the win should be measurable.

**Comment by mykaul** [comment]
Closing this PR — both commits have been superseded/relocated:

1. **`c7ae8a06` (host UUID map for O(1) tablet replica lookups)** is now redundant. Master already merged [`9900f2d5`](https://github.com/scylladb/gocql/commit/9900f2d5c96ccf3638dac979769190ab54133f33) ("perf: index host info snapshots by host id"), which solves the same problem via `hostInfoList`/`cowHostList.hostByID()`. That approach is actually more general (benefits all host policies, not just token-aware) and has a size threshold to skip the map for small clusters, so there's nothing left to add here.

2. **`b43b848b` (cache `qry.IsLWT()` in `Pick()`)** doesn't belong in this PR — it's unrelated to the tablet replica lookup work and got bundled in by mistake. Split out into a small, focused PR on top of current master: #959.

No remaining unique content in this branch, so closing rather than force-pushing an empty diff.
